merkle.go

```diff
@@ -124,30 +133,89 @@ func (t *Tree) AddLeaf(value []byte) error {
+// leavesToProve when the tree was initialized. For a single proved leaf this is a standard merkle proof (one sibling
+// per layer of the tree from the leaves to the root, excluding the proved leaf and root).
+// If the tree is unbalanced (num. of leaves is not a power of 2) it will perform padding on-the-fly.
+func (t *Tree) RootAndProof() ([]byte, [][]byte) {
+	ephemeralProof := t.proof
+	var ephemeralNode node
 	l := t.baseLayer
-	for {
-		if l.next == nil {
-			return l.parking.value, nil
+	for height := uint(0); height < t.minHeight || l != nil; height++ {
+
+		// If we've reached the last layer and the ephemeral node is still empty, the tree is balanced and the parked
+		// node is its root.
+		// In any other case (minHeight not reached, or the tree is unbalanced) we want to add padding at this point.
+		reachedMinHeight := height >= t.minHeight
+		onLastLayer := l != nil && l.next == nil
+		parkingIsBalancedTreeRoot := reachedMinHeight && onLastLayer && ephemeralNode.IsEmpty()
+		if parkingIsBalancedTreeRoot {
+			return l.parking.value, ephemeralProof
 		}
-		if l.parking.value != nil {
-			return nil, ErrorIncompleteTree
+
+		var parking node
+		if l != nil {
+			parking = l.parking
+		}
+		parent, lChild, rChild := t.calcEphemeralParent(parking, ephemeralNode)
+
+		// Consider adding children to the ephemeralProof.
+		if parent.onProvenPath {
+			if !lChild.onProvenPath {
+				ephemeralProof = append(ephemeralProof, lChild.value)
+			}
+			if !rChild.onProvenPath {
+				ephemeralProof = append(ephemeralProof, rChild.value)
+			}
+		}
+		ephemeralNode = parent
+		if l != nil {
+			l = l.next
 		}
-		l = l.next
 	}
+	return ephemeralNode.value, ephemeralProof
 }
 
-// Proof returns a partial tree proving the membership of leaves that were passed in leavesToProve when the tree was
-// initialized or an error if the number of leaves added is not a power of 2. For a single proved leaf this is a
-// standard merkle proof (one sibling per layer of the tree from the leaves to the root, excluding the proved leaf
-// and root).
-func (t *Tree) Proof() ([][]byte, error) {
-	// We call t.Root() to traverse the layers and ensure the tree is full.
-	if _, err := t.Root(); err != nil {
-		return nil, err
+// calcEphemeralParent calculates the parent using the layer parking and ephemeralNode. When one of those is missing it
+// uses PaddingValue to pad. It returns the actual nodes used along with the parent.
+func (t *Tree) calcEphemeralParent(parking, ephemeralNode node) (parent, lChild, rChild node) {
+	switch {
+	case !parking.IsEmpty() && !ephemeralNode.IsEmpty():
+		lChild, rChild = parking, ephemeralNode
+
+	case !parking.IsEmpty() && ephemeralNode.IsEmpty():
+		lChild, rChild = parking, PaddingValue
+
+	case parking.IsEmpty() && !ephemeralNode.IsEmpty():
+		lChild, rChild = ephemeralNode, PaddingValue
+
+	default: // both are empty
+		return ephemeralNode, ephemeralNode, ephemeralNode
```

Comment:
No, the `parent` must be empty, since there's no reason to pad anything. The other values are unimportant.

To reduce confusion I've changed it to `emptyNode` (which I defined at the top of the file).